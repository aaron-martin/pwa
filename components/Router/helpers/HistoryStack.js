/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import EventEmitter from 'events';
import {
  HISTORY_PUSH_ACTION,
  HISTORY_POP_ACTION,
  HISTORY_REPLACE_ACTION,
} from '../../../constants/ActionTypes';

/**
 * Manges the full history stack
 */
class HistoryStack extends EventEmitter {
  /**
   * Event is executed when an entry was removed
   * @type {string}
   */
  static EVENT_ENTRY_REMOVED = 'EVENT_ENTRY_REMOVED';

  /**
   * Event is executed when an entry was removed
   * @type {string}
   */
  static EVENT_ENTRY_ADDED = 'EVENT_ENTRY_ADDED';

  /**
   * Event is executed when an entry got inactive (not on top anymore)
   * @type {string}
   */
  static EVENT_ENTRY_INACTIVE = 'EVENT_ENTRY_INACTIVE';

  /**
   * Event is executed when an entry got active (moved to top)
   * @type {string}
   */
  static EVENT_ENTRY_ACTIVE = 'EVENT_ENTRY_ACTIVE';

  /**
   * Event is executed when a entry changed its values
   * @type {string}
   */
  static EVENT_ENTRY_CHANGED = 'EVENT_ENTRY_CHANGED';

  /**
   * Initializes the HistoryStack class.
   * @param {Object} initialEntry The first entry in the history stack.
   */
  constructor(initialEntry) {
    super();
    this.stack = [];
    this.stack.push(initialEntry);
  }

  /**
   * Searches a history entry by its key.
   * @param {string} key The unique location key.
   * @returns {number}
   */
  findHistoryIndexByKey = key => this.stack.findIndex(e => e.key === key);

  /**
   * Returns the active history entry.
   * @returns {Object}
   */
  getActive = () => this.stack[this.stack.length - 1];

  /**
   * Returns the history length.
   * @returns {number}
   */
  getLength = () => this.stack.length;

  /**
   * Updates the last stack entry
   * @param {Object} newEntry The new stack entry
   */
  updateStack(newEntry) {
    this.stack[this.stack.length - 1] = newEntry;
  }

  /**
   * Applies a real history change to this history stack.
   * @param {string} action The action that caused history change (PUSH, POP, REPLACE)
   * @param {Object} location The location object.
   */
  applyChange(action, location) {
    switch (action) {
      case HISTORY_PUSH_ACTION: {
        // Add a new history entry.
        const updatedLocation = {
          ...location,
          immutableKey: location.key,
        };
        this.stack.push(updatedLocation);

        this.emit(HistoryStack.EVENT_ENTRY_ADDED, updatedLocation);
        this.emit(HistoryStack.EVENT_ENTRY_INACTIVE, this.stack[this.stack.length - 2]);
        break;
      }

      case HISTORY_REPLACE_ACTION: {
        /**
         * Replace causes the key to be changed. This only causes confusion
         * therefore HistoryStack provides an additional immutableKey that will never change.
         */
        const before = this.stack[this.stack.length - 1];

        if (before.pathname !== location.pathname) {
          const updatedLocation = {
            ...location,
            immutableKey: location.key,
          };

          this.updateStack(updatedLocation);

          this.emit(HistoryStack.EVENT_ENTRY_REMOVED, before);
          this.emit(HistoryStack.EVENT_ENTRY_ADDED, updatedLocation);
        } else {
          this.updateStack({
            ...location,
            immutableKey: before.immutableKey,
          });
        }

        this.emit(
          HistoryStack.EVENT_ENTRY_CHANGED,
          this.stack[this.stack.length - 1],
          before
        );
        break;
      }

      case HISTORY_POP_ACTION: {
        /**
         * Pop all history entries up to the one we are looking for
         * and fire an event for each entry that was removed.
         */
        const targetIndex = this.findHistoryIndexByKey(location.key || 'root');
        if (targetIndex === -1) {
          // When going back to a page that is no longer stored push it instead.
          this.applyChange(HISTORY_PUSH_ACTION, location);
          return;
        }

        // Clean the stack and store removed entries.
        const removedEntries = this.stack
          .slice(targetIndex + 1, this.stack.length);
        this.stack = this.stack.slice(0, targetIndex + 1);

        // Notify removed entries.
        removedEntries.forEach(target => this.emit(HistoryStack.EVENT_ENTRY_REMOVED, target));

        // Active changed
        this.emit(HistoryStack.EVENT_ENTRY_ACTIVE, this.stack[this.stack.length - 1]);
        break;
      }

      default:
        break;
    }
  }
}

export default HistoryStack;
