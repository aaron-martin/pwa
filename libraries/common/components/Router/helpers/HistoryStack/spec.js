/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  HISTORY_PUSH_ACTION,
  HISTORY_POP_ACTION,
  HISTORY_REPLACE_ACTION,
} from '../../../../constants/ActionTypes';
import HistoryStack from './index';

describe('HistoryStack', () => {
  it('should trigger added and inactive event when history entry was added', () => {
    const stack = new HistoryStack({ key: 'root' });
    const addedSpy = jest.fn();
    const inactiveSpy = jest.fn();

    stack.on(HistoryStack.EVENT_ENTRY_ADDED, addedSpy);
    stack.on(HistoryStack.EVENT_ENTRY_INACTIVE, inactiveSpy);
    stack.applyChange(HISTORY_PUSH_ACTION, { key: '123' });

    expect(addedSpy).toHaveBeenCalledTimes(1);
    expect(addedSpy).toHaveBeenLastCalledWith({
      immutableKey: '123',
      key: '123',
    });
    expect(inactiveSpy).toHaveBeenCalledTimes(1);
    expect(inactiveSpy).toHaveBeenLastCalledWith({ key: 'root' });
  });

  it('should trigger removed event for all previous entries', () => {
    const stack = new HistoryStack({ key: 'root' });
    const removedSpy = jest.fn();
    const activeSpy = jest.fn();

    // Navigate 3 pages forward
    stack.applyChange(HISTORY_PUSH_ACTION, { key: '123' });
    stack.applyChange(HISTORY_PUSH_ACTION, { key: '456' });
    stack.applyChange(HISTORY_PUSH_ACTION, { key: '789' });

    // Navigate back to the first (-2)
    stack.on(HistoryStack.EVENT_ENTRY_REMOVED, removedSpy);
    stack.on(HistoryStack.EVENT_ENTRY_ACTIVE, activeSpy);
    stack.applyChange(HISTORY_POP_ACTION, {
      immutableKey: '123',
      key: '123',
    });

    // Pop should have been called on all old routes.
    expect(removedSpy).toHaveBeenCalledTimes(2);
    expect(removedSpy).toHaveBeenCalledWith({
      immutableKey: '789',
      key: '789',
    });
    expect(removedSpy).toHaveBeenCalledWith({
      immutableKey: '456',
      key: '456',
    });

    // Active should be the first entry again.
    expect(activeSpy).toHaveBeenCalledTimes(1);
    expect(activeSpy).toHaveBeenCalledWith({
      immutableKey: '123',
      key: '123',
    });
  });

  it('should not mutate the immutableKey when replacing history', () => {
    const stack = new HistoryStack({ key: 'root' });
    const changedSpy = jest.fn();

    stack.on(HistoryStack.EVENT_ENTRY_CHANGED, changedSpy);
    stack.applyChange(HISTORY_PUSH_ACTION, { key: '123' });
    stack.applyChange(HISTORY_REPLACE_ACTION, { key: '456' });

    expect(stack.getActive().key).toEqual('456');
    expect(stack.getActive().immutableKey).toEqual('123');

    expect(changedSpy).toHaveBeenCalledTimes(1);
    expect(changedSpy).toHaveBeenCalledWith(
      {
        immutableKey: '123',
        key: '456',
      },
      {
        immutableKey: '123',
        key: '123',
      }
    );
  });

  it('should allow history changes inside the event handler', () => {
    const stack = new HistoryStack({ key: 'root' });

    /**
     * Replaces history immediately after adding.
     */
    const added = () => {
      stack.applyChange(HISTORY_REPLACE_ACTION, { key: '123' });
    };

    /**
     * Replaces history immediately after removing.
     */
    const removed = () => {
      stack.applyChange(HISTORY_REPLACE_ACTION, { key: '456' });
    };

    stack.on(HistoryStack.EVENT_ENTRY_ADDED, added);
    stack.on(HistoryStack.EVENT_ENTRY_REMOVED, removed);

    stack.applyChange(HISTORY_PUSH_ACTION, { key: 'x123' });
    stack.applyChange(HISTORY_PUSH_ACTION, { key: 'foo' });
    stack.applyChange(HISTORY_POP_ACTION, { key: '123' });

    expect(stack.stack).toEqual([
      { key: 'root' },
      { immutableKey: 'x123', key: '456' },
    ]);
  });
});
