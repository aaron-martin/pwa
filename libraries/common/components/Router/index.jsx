/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { Component } from 'react';
import PropTypes from 'prop-types';
import pathMatch from 'path-match';
import HistoryStack from './helpers/HistoryStack';

// Configure the path matcher that will parse the url.
const matcher = pathMatch({
  sensitive: false,
  strict: false,
  end: true,
});

/**
 * The Router component.
 * @see https://shopgate.atlassian.net/wiki/display/CONSUMER/Shopgate+React+Router
 */
class Router extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    history: PropTypes.shape().isRequired,
  };

  static childContextTypes = {
    registerRoute: PropTypes.func.isRequired,
    history: PropTypes.shape().isRequired,
  };

  /**
   * Initializes the router component.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.routeRegistry = {};
  }

  /**
   * Sets the context for the current router.
   * @returns {Object}
   */
  getChildContext() {
    return {
      history: this.props.history,

      /**
       * Registers a route.
       * @param {string} route Route definition.
       * @param {Function} add Callback that is executed when route needs to be added.
       * @param {Function} remove Callback that is executed when a active route needs to be removed.
       * @param {Function} show Callback that is executed when a hidden route needs to be shown.
       * @param {Function} hide Callback that is executed when a active route needs to be hidden.
       * @param {Object} options Options for the route.
       */
      registerRoute: (route, add, remove, show, hide, options) => {
        this.routeRegistry[route] = {
          add,
          remove,
          show,
          hide,
          options,
          matcher: matcher(route),
        };
      },
    };
  }

  /**
   * The first time the component is mounted the current history will be used as initial route.
   */
  componentDidMount() {
    /**
     * Notifies a route using the given action and location
     * @param {Object} location The location object.
     * @param {string} action The action that is requested.
     * @param {Object} [updatedLocation] The updated location object.
     */
    const notifyRoute = (location, action, updatedLocation) => {
      const result = this.findRouteForLocation(location);

      if (!result) {
        return;
      }

      if (typeof result.route[action] === 'function') {
        result.route[action](location, result.params, updatedLocation);
      }
    };

    // Notify initial / first root.
    notifyRoute(this.props.history.getActive(), 'add');

    // Listen on relevant history changes.
    this.props.history
      .on(HistoryStack.EVENT_ENTRY_ADDED, location => notifyRoute(location, 'add'))
      .on(HistoryStack.EVENT_ENTRY_REMOVED, location => notifyRoute(location, 'remove'))
      .on(HistoryStack.EVENT_ENTRY_ACTIVE, location => notifyRoute(location, 'show'))
      .on(HistoryStack.EVENT_ENTRY_INACTIVE, location => notifyRoute(location, 'hide'));
  }

  /**
   * Searches for a registered route that matches given location.
   * If no matching route was found the 'missmatch' route will be used (if available)
   * @param {Object} location Location object.
   * @returns {Object} The registered route instance.
   */
  findRouteForLocation(location) {
    let matchParams = null;
    let matchRoute = null;
    let missMatch = null;

    Object.keys(this.routeRegistry).some((defintion) => {
      const route = this.routeRegistry[defintion];

      // If route is a miss route save it for later use.
      if (route.options.miss) {
        missMatch = route;
      }

      // If we find a match we will take this as active route.
      const tempMatch = route.matcher(location.pathname);
      if (tempMatch) {
        matchParams = tempMatch;
        matchRoute = route;

        return true;
      }

      return false;
    });

    if (matchParams) {
      return {
        route: matchRoute,
        params: matchParams,
      };
    } else if (missMatch) {
      return {
        route: missMatch,
        params: {},
      };
    }

    return null;
  }

  /**
   * Renders the component children.
   * @returns {JSX}
   */
  render() {
    return this.props.children;
  }
}

export default Router;
