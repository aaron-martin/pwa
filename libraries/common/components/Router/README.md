# Router
---

The router component cares about the given history and changes the visible route.
The history will notify the router about url changes and the router will then notify
the matching route.

> **Dependents:** <br> `<Route />`

## Getting Started

```
import { Router, Route } from 'Library/components';
import { Foo, Bar } from './somewhere';

<Router>
  <div>
      <Route path="/category/:categoryId" component={Foo} />
      <Route path="/item/:productId" component={Bar} />
  </div>
</Router>
```

## Context

### history

_Type_: `Object`<br>

The history prop will be passed in the context to all children.

### registerRoute

_Type_: `function (route, callback, options)`<br>

Can be used to register a route. When the history object notifies the router about url changes
the passed `callback` will be executed if matched.

## Props

### children (required)

_Type_: `React.Component`<br>

Needs to be a single child! Will render the child component with the new context.

###### Usage:

```
<Router>
  <div>
      <Route path="/category/:categoryId" component={Foo} />
      <Route path="/item/:productId" component={Bar} />
  </div>
</Router>
```

### history (required)

_Type_: `Object`<br>

The history object. For more information see: https://www.npmjs.com/package/history

###### Usage:

```
import createHistory from 'history/createBrowserHistory';

const history = createHistory();

<Router history={history}>
  <div>
      <Route path="/category/:categoryId" component={Foo} />
      <Route path="/item/:productId" component={Bar} />
  </div>
</Router>
```
---
