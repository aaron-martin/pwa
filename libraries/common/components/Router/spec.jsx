/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import Router from './index';
import MockRoute from './components/Route/mock';
import HistoryStack from './helpers/HistoryStack';

describe('<Router />', () => {
  it('should notify route about initial location', () => {
    const mockHistory = new HistoryStack({
      key: 'c123',
      pathname: '/category/123',
    });
    const mockAdd = jest.fn();

    const wrapper = mount(
      <Router history={mockHistory}>
        <MockRoute mockAdd={mockAdd} path="/category/:id" />
      </Router>
    );

    expect(wrapper).toMatchSnapshot();
    expect(mockAdd).toHaveBeenCalledTimes(1);
    expect(mockAdd).toHaveBeenCalledWith(
      {
        key: 'c123',
        pathname: '/category/123',
      },
      {
        id: '123',
      },
      undefined
    );
  });

  it('should notify previous route to hide and new one to show', () => {
    const mockHistory = new HistoryStack({
      key: 'c123',
      pathname: '/category/123',
    });
    const mockAddCategory = jest.fn();
    const mockHideCategory = jest.fn();
    const mockAddProduct = jest.fn();

    const wrapper = mount(
      <Router history={mockHistory}>
        <div>
          <MockRoute mockAdd={mockAddCategory} mockHide={mockHideCategory} path="/category/:id" />
          <MockRoute mockAdd={mockAddProduct} path="/item/:id" />
        </div>
      </Router>
    );

    mockHistory.applyChange('PUSH', {
      key: 'i123',
      pathname: '/item/123',
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(mockAddCategory).toHaveBeenCalledTimes(1);
    expect(mockAddCategory).toHaveBeenCalledWith(
      {
        key: 'c123',
        pathname: '/category/123',
      },
      { id: '123' },
      undefined
    );
    expect(mockHideCategory).toHaveBeenCalledTimes(1);
    expect(mockHideCategory).toHaveBeenCalledWith(
      {
        key: 'c123',
        pathname: '/category/123',
      },
      { id: '123' },
      undefined
    );
    expect(mockAddProduct).toHaveBeenCalledTimes(1);
    expect(mockAddProduct).toHaveBeenCalledWith(
      {
        immutableKey: 'i123',
        key: 'i123',
        pathname: '/item/123',
      },
      { id: '123' },
      undefined
    );
  });

  it('should notify previous route to destroy', () => {
    const mockHistory = new HistoryStack({
      key: 'c123',
      pathname: '/category/123',
    });
    const mockRemoveItem = jest.fn();

    const wrapper = mount(
      <Router history={mockHistory}>
        <div>
          <MockRoute path="/category/:id" />
          <MockRoute mockAdd={mockRemoveItem} path="/item/:id" />
        </div>
      </Router>
    );

    mockHistory.applyChange('PUSH', {
      key: 'i123',
      pathname: '/item/123',
    });
    mockHistory.applyChange('POP', {
      key: 'c123',
      pathname: '/category/123',
    });
    wrapper.update();

    expect(mockRemoveItem).toHaveBeenCalledTimes(1);
    expect(mockRemoveItem).toHaveBeenCalledWith(
      {
        immutableKey: 'i123',
        key: 'i123',
        pathname: '/item/123',
      },
      { id: '123' },
      undefined
    );
  });
});
