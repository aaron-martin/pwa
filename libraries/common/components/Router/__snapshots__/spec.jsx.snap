// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Router /> should notify previous route to hide and new one to show 1`] = `
<Router
  history={
    HistoryStack {
      "_events": Object {
        "EVENT_ENTRY_ACTIVE": [Function],
        "EVENT_ENTRY_ADDED": [Function],
        "EVENT_ENTRY_INACTIVE": [Function],
        "EVENT_ENTRY_REMOVED": [Function],
      },
      "_eventsCount": 4,
      "_maxListeners": undefined,
      "domain": null,
      "findHistoryIndexByKey": [Function],
      "getActive": [Function],
      "getLength": [Function],
      "stack": Array [
        Object {
          "key": "c123",
          "pathname": "/category/123",
        },
        Object {
          "immutableKey": "i123",
          "key": "i123",
          "pathname": "/item/123",
        },
      ],
    }
  }
>
  <div>
    <MockRoute
      mockAdd={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "key": "c123",
                "pathname": "/category/123",
              },
              Object {
                "id": "123",
              },
              undefined,
            ],
          ],
        }
      }
      mockHide={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "key": "c123",
                "pathname": "/category/123",
              },
              Object {
                "id": "123",
              },
              undefined,
            ],
          ],
        }
      }
      mockRemove={[Function]}
      mockShow={[Function]}
      path="/category/:id"
    />
    <MockRoute
      mockAdd={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "immutableKey": "i123",
                "key": "i123",
                "pathname": "/item/123",
              },
              Object {
                "id": "123",
              },
              undefined,
            ],
          ],
        }
      }
      mockHide={[Function]}
      mockRemove={[Function]}
      mockShow={[Function]}
      path="/item/:id"
    />
  </div>
</Router>
`;

exports[`<Router /> should notify route about initial location 1`] = `
<Router
  history={
    HistoryStack {
      "_events": Object {
        "EVENT_ENTRY_ACTIVE": [Function],
        "EVENT_ENTRY_ADDED": [Function],
        "EVENT_ENTRY_INACTIVE": [Function],
        "EVENT_ENTRY_REMOVED": [Function],
      },
      "_eventsCount": 4,
      "_maxListeners": undefined,
      "domain": null,
      "findHistoryIndexByKey": [Function],
      "getActive": [Function],
      "getLength": [Function],
      "stack": Array [
        Object {
          "key": "c123",
          "pathname": "/category/123",
        },
      ],
    }
  }
>
  <MockRoute
    mockAdd={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "key": "c123",
              "pathname": "/category/123",
            },
            Object {
              "id": "123",
            },
            undefined,
          ],
        ],
      }
    }
    mockHide={[Function]}
    mockRemove={[Function]}
    mockShow={[Function]}
    path="/category/:id"
  />
</Router>
`;
